### PR TITLE
test: upgrade multires example to use sharding

### PR DESCRIPTION
**Pull request by cristianmtr**
This PR will commit **version 3** (latest git commit) from the following discussion

---

**IMPORTANT NOTE**: This is probably something that will affect **all** examples where sharding is used. 

Issue with batching and missing results.

Even after **ex. v3** fixed the issue mentioned below, I noticed that the results were different, and the top matches were missing.

See screenshot here:

image

With some debugging, we realized that it was due to `qeury.yml`

```
  chunk_idx:
    uses: pods/chunk.yml
    shards: $JINA_SHARDS
    separated_workspace: true
    polling: all
    uses_reducing: _merge_all
    timeout_ready: -1 # larger timeout as in query time will read all the data
```

This needs to be 

```
  chunk_idx:
    uses: pods/chunk.yml
    shards: $JINA_SHARDS
    separated_workspace: true
    polling: all
    uses_after: pods/chunk_merger.yml
    timeout_ready: -1 # larger timeout as in query time will read all the data
```

with `chunk_merger.yml` like:

```
!BaseExecutor
with: {}
metas:
  name: merge
requests:
  on:
    [SearchRequest, TrainRequest, IndexRequest, ControlRequest]:
      - !ReduceAllDriver
        with:
          traversal_paths: ['cm']
```

This also required to ensure correct sorting of the matches, as is done now.

---

I have been trying to extend the multires lyrics example to use sharding.

There seems to be some bug related to sharding and batching. 

I came up with the following steps to reproduce and test what the problem was.

**Experiment v1**

git: 96863bec5de24b589fab943bad92fde5328a3ef0    

* data = 8 rows
* shards = 2
* no batching
* results contain matches?
    * console run: yes
    * pycharm run: yes
    * pycharm debug run: no -- 'matches' not present in query result  
* The following error is seen in all the above cases.

```
Traceback (most recent call last):
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/peapods/pea.py", line 292, in msg_callback
    self.zmqlet.send_message(self._callback(msg))
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/peapods/pea.py", line 266, in _callback
    self.pre_hook(msg).handle(msg).post_hook(msg)
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/peapods/pea.py", line 160, in handle
    self.executor(self.request_type)
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/executors/__init__.py", line 570, in __call__
    d()
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/drivers/__init__.py", line 251, in __call__
    self._traverse_apply(self.req.docs, *args, **kwargs)
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/drivers/__init__.py", line 268, in _traverse_apply
    **kwargs,
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/drivers/__init__.py", line 281, in _traverse_rec
    doc.chunks, doc, 'chunks', path[1:], *args, **kwargs
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/drivers/__init__.py", line 284, in _traverse_rec
    self._apply_all(docs, parent_doc, parent_edge_type, *args, **kwargs)
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/drivers/search.py", line 124, in _apply_all
    idx, dist = self.exec_fn(embed_vecs, top_k=int(self.top_k))
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/executors/indexers/vector.py", line 279, in query
    dist = self._euclidean(_keys, self.query_handler)
  File "/home/cristian/envs/lyrics/lib/python3.7/site-packages/jina/executors/decorators.py", line 241, in arg_wrapper
    for b in batch_iterator(data[:total_size], b_size, split_over_axis, yield_slice=yield_slice):
TypeError: 'NoneType' object is not subscriptable
```

**Experiment v2**

_Is it because of the lack of data?_

git: 01d73d409c6ba7f3368a5d8ad6baefd84671de3e
* data = 16 rows (duplicates of the 8 from ex. v1)
* shards  = 2 
* no batching
* results contain matches?
    * console run: yes
    * pycharm run: yes
    * pycharm debug run: no -- 'matches' not present in query result 
* The above error is seen in all the above cases.

**Experiment v3**

_Is it because we're not batching?_

git c4cc3ffd84efd1a6b6db1807408e4fb3f9f7a21d
* data = 8 rows
* shards  = 2 
* batching = 4
* results contain matches?
    * console run: yes
    * pycharm run: yes
    * pycharm debug run: yes
* Error not present

Conclusion? Seems that sharding doesn't work if we don't also use batching. Is this intended?

### CODE REVIEW

File: multires-lyrics-search/tests/test_multireslyrics.py [hunk 1]
```diff
@@ -94,16 +94,22 @@ def test_query(tmpdir, queries_and_expected_replies):
                 for match in chunk['matches']:
                     chunk_result['chunk_matches'].append(match['text'])
                 query_chunk_results.append(chunk_result)
-            assert query_chunk_results == exp_result["chunk-level"]
+            # TODO remove assertion because it would require re-writing the query-results.json
```
Comment (JoanFM) [inline multires-lyrics-search/tests/test_multireslyrics.py:97]: Remember to get back to assertions

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: multires-lyrics-search/tests/test_multireslyrics.py (hunk 1 above)
Comment (cristianmtr) [reply multires-lyrics-search/tests/test_multireslyrics.py:97]: Oh yes! Missed that

File: multires-lyrics-search/tests/test_multireslyrics.py [hunk 2]
```diff
@@ -50,7 +49,7 @@ def index_documents():
     f = Flow().load_config(INDEX_FLOW_FILE_PATH)
 
     with f:
-        f.index(input_fn)
+        f.index(input_fn, batch_size=4)
```
Comment (bhavsarpratik) [inline multires-lyrics-search/tests/test_multireslyrics.py:52]: Good to define variable `BATCH_SIZE` at the top

File: multires-lyrics-search/tests/test_multireslyrics.py [hunk 3]
```diff
@@ -79,6 +78,7 @@ def queries_and_expected_replies():
 
 
 def test_query(tmpdir, queries_and_expected_replies):
+    print(f'tmpdir = {tmpdir}')
```
Comment (bhavsarpratik) [inline multires-lyrics-search/tests/test_multireslyrics.py:81]: Can be removed now

**Review by bhavsarpratik — CHANGES_REQUESTED** [review]

**Comment by cristianmtr** [comment]
Just FYI: I am experiencing issues with the results quality after the batching. Debugging now

File: multires-lyrics-search/flows/query.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@ pods:
     shards: $JINA_SHARDS
     separated_workspace: true
     polling: all
-    uses_reducing: _merge_all
+    uses_after: pods/chunk_merger.yml
```
Comment (JoanFM) [inline multires-lyrics-search/flows/query.yml:19]: U can use _reduce_all I think or _reduce I do not remember exactly

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: multires-lyrics-search/flows/query.yml (hunk 4 above)
Comment (cristianmtr) [reply multires-lyrics-search/flows/query.yml:19]: That didn't work. It's because it need to be set to `traversal: ['cm']`.

**Review by JoanFM — APPROVED** [review]

File: multires-lyrics-search/pods/chunk_merger.yml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+!BaseExecutor
+with: {}
+metas:
+  name: merge
+requests:
+  on:
+    [SearchRequest, TrainRequest, IndexRequest, ControlRequest]:
```
Comment (bhavsarpratik) [inline multires-lyrics-search/pods/chunk_merger.yml:7]: When do we need TrainRequest?

Comment (JoanFM) [reply multires-lyrics-search/pods/chunk_merger.yml:7]: It is not curetly active the `TraninRequest` I would remove it, but what is more important is to remove `ControlRequest` from here

**Review by JoanFM — COMMENTED** [review]
Remove ControlRequest from `ReduceAllDrvier`

**Review by JoanFM — CHANGES_REQUESTED** [review]
Sorry, need to remove `ControlRequest`

**Review by bhavsarpratik — APPROVED** [review]

File: multires-lyrics-search/tests/test_multireslyrics.py [hunk 6]
```diff
@@ -106,4 +113,15 @@ def test_query(tmpdir, queries_and_expected_replies):
...
+            all_results.append(
+                [
+                    query,
+                    {
+                        "match-level": match_result,
+                        "chunk-level": query_chunk_results
+                    }
+                ]
+            )
+    with open(f'results batch size {BATCH_SIZE}.json', 'w') as f:
```
Comment (JoanFM) [inline multires-lyrics-search/tests/test_multireslyrics.py:126]: remove this wriing right?

File: multires-lyrics-search/tests/test_multireslyrics.py [hunk 7]
```diff
@@ -91,13 +93,18 @@ def test_query(tmpdir, queries_and_expected_replies):
...
                 query_chunk_results.append(chunk_result)
             assert query_chunk_results == exp_result["chunk-level"]
 
             # match-level comparison
             matches = output['search']['docs'][0]['matches']
+            # make sure to sort in asc. order, by score
+            matches = sorted(matches, key=lambda x: x['score']['value'], reverse=False)
             match_result = []
             for match in matches:
                 match_text = match['text']
```
Comment (JoanFM) [inline multires-lyrics-search/tests/test_multireslyrics.py:110]: why do not we have an assert at `doc level` of matches anymore?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: multires-lyrics-search/tests/test_multireslyrics.py (hunk 7 above)
Comment (cristianmtr) [reply multires-lyrics-search/tests/test_multireslyrics.py:110]: You mean this https://github.com/jina-ai/examples/blob/a38608b685e426fbb412462147d093a0bd7509b2/multires-lyrics-search/tests/test_multireslyrics.py#L112 ?

File: multires-lyrics-search/tests/test_multireslyrics.py (hunk 6 above)
Comment (cristianmtr) [reply multires-lyrics-search/tests/test_multireslyrics.py:126]: I think it's useful for debugging. But I can see that other examples don't have this, so maybe I should remove it to keep things consistent?

File: multires-lyrics-search/tests/test_multireslyrics.py (hunk 7 above)
Comment (cristianmtr) [reply multires-lyrics-search/tests/test_multireslyrics.py:110]: Maybe it should be renamed to `"doc-level"`

Comment (JoanFM) [reply multires-lyrics-search/tests/test_multireslyrics.py:110]: Ah no my bad, yes the diff only showed the changes and got me confused.

File: multires-lyrics-search/tests/test_multireslyrics.py (hunk 6 above)
Comment (JoanFM) [reply multires-lyrics-search/tests/test_multireslyrics.py:126]: Yes, please remove it. Once one wants to debug can start doing their prints. But this test is a test to be run by our CI CDs

Comment (cristianmtr) [reply multires-lyrics-search/tests/test_multireslyrics.py]: Done

**Review by JoanFM — APPROVED** [review]

**Pull request merged**